src/error_estimation/smoothness_estimator.C

```diff
@@ -0,0 +1,388 @@
+
+  // The current mesh
+  const MeshBase & mesh = system.get_mesh();
+
+  // Resize the smoothness_per_cell vector to be
+  // the number of elements, initialize it to 0.
+  smoothness_per_cell.resize (mesh.max_elem_id());
+  std::fill (smoothness_per_cell.begin(), smoothness_per_cell.end(), 0.);
+
+  // Prepare current_local_solution to localize a non-standard
+  // solution vector if necessary
+  if (solution_vector && solution_vector != system.solution.get())
+    {
+      NumericVector<Number> * newsol =
+        const_cast<NumericVector<Number> *>(solution_vector);
+      System & sys = const_cast<System &>(system);
+      newsol->swap(*sys.solution);
+      sys.update();
+    }
+
+  //------------------------------------------------------------
+  // Iterate over all the active elements in the mesh
+  // that live on this processor.
+  Threads::parallel_for (ConstElemRange(mesh.active_local_elements_begin(),
+                                        mesh.active_local_elements_end(),
+                                        200),
+                         EstimateError(system,
+                                       *this,
+                                       smoothness_per_cell)
+                         );
+
+  // Each processor has now computed the smoothness for its local
+  // elements, and smoothness_per_cell contains 0 for all the
+  // non-local elements.  Summing the vector will provide the true
+  // value for each element, local or remote
+  this->reduce_error(smoothness_per_cell, system.comm());
+
+  // If we used a non-standard solution before, now is the time to fix
+  // the current_local_solution
+  if (solution_vector && solution_vector != system.solution.get())
+    {
+      NumericVector<Number> * newsol = const_cast<NumericVector<Number> *>(solution_vector);
+      System & sys = const_cast<System &>(system);
+      newsol->swap(*sys.solution);
+      sys.update();
+    }
+}
+
+
+void SmoothnessEstimator::EstimateError::operator()(const ConstElemRange & range) const
+{
+  // The current mesh
+  const MeshBase & mesh = system.get_mesh();
+
+  // The dimensionality of the mesh
+  const unsigned int dim = mesh.mesh_dimension();
+
+  // The number of variables in the system
+  const unsigned int n_vars = system.n_vars();
+
+  // The DofMap for this system
+  const DofMap & dof_map = system.get_dof_map();
+
+  //------------------------------------------------------------
+  // Iterate over all the elements in the range.
+for (const auto & elem : range)
+{
+  const dof_id_type e_id = elem->id();
+  // Loop over each variable
+  for (unsigned int var=0; var<n_vars; var++)
+  {
+    if (error_estimator.error_norm.weight(var) == 0.0)
+      continue;
+
+    const FEType & fe_type = dof_map.variable_type(var);
+    const Order element_order = fe_type.order + elem->p_level();
+
+    if (element_order == 1)
+      libmesh_warning("Warning: Polynomial order in element is 1, smoothness cannot be stimated!\n"
```

Comment:
```suggestion
      libmesh_warning("Warning: Polynomial order in element is 1, smoothness cannot be estimated!\n"
```

Reply:
Although, I'm not sure we want a warning here at all.  Surely we expect to hit `element_order=1` cases legitimately, and we don't need to warn the user because they're not doing anything wrong?

But we should document (not here, but in the header) what *does* end up in the output `ErrorVector` in those elements' entries in those cases.